File: examples/java/com/intel/daal/examples/optimization_solvers/SGDMomentOptResDenseBatch.java
/* file: SGDMomentOptResDenseBatch.java */
/*******************************************************************************
* Copyright 2014-2018 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
 //  Content:
 //     Java example of dense SGD momentum in the batch processing mode
 //     with optional result calculation and its usage in the next run
 ////////////////////////////////////////////////////////////////////////////////
 */

/**
 * <a name="DAAL-EXAMPLE-JAVA-SGDMOMENTOPTRESDENSEBATCH">
 * @example SGDMomentOptResDenseBatch.java
 */

package com.intel.daal.examples.optimization_solvers;

import com.intel.daal.algorithms.optimization_solver.sgd.*;
import com.intel.daal.algorithms.optimization_solver.iterative_solver.Input;
import com.intel.daal.algorithms.optimization_solver.iterative_solver.InputId;
import com.intel.daal.algorithms.optimization_solver.iterative_solver.OptionalInputId;
import com.intel.daal.algorithms.optimization_solver.iterative_solver.Result;
import com.intel.daal.algorithms.optimization_solver.iterative_solver.ResultId;
import com.intel.daal.algorithms.optimization_solver.iterative_solver.OptionalResultId;
import com.intel.daal.algorithms.OptionalArgument;
import com.intel.daal.data_management.data.HomogenNumericTable;
import com.intel.daal.data_management.data.NumericTable;
import com.intel.daal.data_management.data.MergedNumericTable;
import com.intel.daal.data_management.data_source.DataSource;
import com.intel.daal.data_management.data_source.FileDataSource;
import com.intel.daal.examples.utils.Service;
import com.intel.daal.services.DaalContext;

class SGDMomentOptResDenseBatch {
    private static final long nFeatures = 3;
    private static final double accuracyThreshold = 0.0000001;
    private static final long nIterations = 400;
    private static final long batchSize = 4;
    private static final double learningRate = 0.5;
    private static double[] startPoint = {8, 2, 1, 4};
    /* Input data set parameters */
    private static final String dataFileName = "../data/batch/mse.csv";

    private static DaalContext context = new DaalContext();

    public static void main(String[] args) throws java.io.FileNotFoundException, java.io.IOException {

        /* Retrieve the data from input data sets */
        FileDataSource dataSource = new FileDataSource(context, dataFileName,
                DataSource.DictionaryCreationFlag.DoDictionaryFromContext,
                DataSource.NumericTableAllocationFlag.NotAllocateNumericTable);

        /* Create Numeric Tables for data and values for dependent variable */
        NumericTable data = new HomogenNumericTable(context, Float.class, nFeatures, 0, NumericTable.AllocationFlag.DoNotAllocate);
        NumericTable dataDependents = new HomogenNumericTable(context, Float.class, 1, 0, NumericTable.AllocationFlag.DoNotAllocate);
        MergedNumericTable mergedData = new MergedNumericTable(context);
        mergedData.addNumericTable(data);
        mergedData.addNumericTable(dataDependents);

        /* Retrieve the data from an input file */
        dataSource.loadDataBlock(mergedData);

        /* Create an MSE objective function to compute a SGD momentum */
        com.intel.daal.algorithms.optimization_solver.mse.Batch mseFunction =
            new com.intel.daal.algorithms.optimization_solver.mse.Batch(context, Float.class,
                    com.intel.daal.algorithms.optimization_solver.mse.Method.defaultDense, data.getNumberOfRows());

        mseFunction.getInput().set(com.intel.daal.algorithms.optimization_solver.mse.InputId.data, data);
        mseFunction.getInput().set(com.intel.daal.algorithms.optimization_solver.mse.InputId.dependentVariables, dataDependents);

        /* Create algorithm objects to compute SGD momentum results */
        Batch sgdAlgorithm = new Batch(context, Float.class, Method.momentum);
        ParameterMomentum algorithmParameter = (ParameterMomentum)sgdAlgorithm.parameter;
        algorithmParameter.setFunction(mseFunction);
        algorithmParameter.setLearningRateSequence(
            new HomogenNumericTable(context, Float.class, 1, 1, NumericTable.AllocationFlag.DoAllocate, learningRate));
        algorithmParameter.setNIterations(nIterations/2);
        algorithmParameter.setAccuracyThreshold(accuracyThreshold);
        algorithmParameter.setBatchSize(batchSize);
        algorithmParameter.setOptionalResultRequired(true);
        sgdAlgorithm.input.set(InputId.inputArgument, new HomogenNumericTable(context, startPoint, 1, nFeatures + 1));

        /* Compute the SGD momentum result for MSE objective function matrix */
        Result result = sgdAlgorithm.compute();

        Service.printNumericTable("Minimum after first compute():",  result.get(ResultId.minimum));
        Service.printNumericTable("Number of iterations performed:",  result.get(ResultId.nIterations));

        sgdAlgorithm.input.set(InputId.inputArgument, result.get(ResultId.minimum));
        sgdAlgorithm.input.set(OptionalInputId.optionalArgument, result.get(OptionalResultId.optionalResult));

        /* Compute the SGD momentum result for MSE objective function matrix */
        result = sgdAlgorithm.compute();

        Service.printNumericTable("Minimum after second compute():",  result.get(ResultId.minimum));
        Service.printNumericTable("Number of iterations performed:",  result.get(ResultId.nIterations));

        context.dispose();
    }
}

File: examples/daal/cpp/source/optimization_solvers/sgd_moment_opt_res_dense_batch.cpp
/* file: sgd_moment_opt_res_dense_batch.cpp */
/*******************************************************************************
* Copyright 2014 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
!  Content:
!    C++ example of the Stochastic gradient descent algorithm
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-SGD_MOMENT_OPT_RES_DENSE_BATCH"></a>
 * \example sgd_moment_opt_res_dense_batch.cpp
 */

#include "daal.h"
#include "service.h"

using namespace daal;
using namespace daal::algorithms;
using namespace daal::data_management;

std::string datasetFileName = "../data/batch/mse.csv";

const size_t nIterations = 400;
const size_t nFeatures = 3;
const float learningRate = 0.5;
const size_t batchSize = 4;
const double accuracyThreshold = 0.0000001;

float initialPoint[nFeatures + 1] = { 8, 2, 1, 4 };

int main(int argc, char* argv[]) {
    checkArguments(argc, argv, 1, &datasetFileName);

    /* Initialize FileDataSource<CSVFeatureManager> to retrieve the input data from a .csv file */
    FileDataSource<CSVFeatureManager> dataSource(datasetFileName,
                                                 DataSource::notAllocateNumericTable,
                                                 DataSource::doDictionaryFromContext);

    /* Create Numeric Tables for data and values for dependent variable */
    NumericTablePtr data(new HomogenNumericTable<>(nFeatures, 0, NumericTable::doNotAllocate));
    NumericTablePtr dependentVariables(
        new HomogenNumericTable<>(1, 0, NumericTable::doNotAllocate));
    NumericTablePtr mergedData(new MergedNumericTable(data, dependentVariables));

    /* Retrieve the data from the input file */
    dataSource.loadDataBlock(mergedData.get());
    size_t nVectors = data->getNumberOfRows();

    services::SharedPtr<optimization_solver::mse::Batch<> > mseObjectiveFunction(
        new optimization_solver::mse::Batch<>(nVectors));
    mseObjectiveFunction->input.set(optimization_solver::mse::data, data);
    mseObjectiveFunction->input.set(optimization_solver::mse::dependentVariables,
                                    dependentVariables);

    /* Create objects to compute the Stochastic momentum gradient descent result using the default method */
    optimization_solver::sgd::Batch<float, optimization_solver::sgd::momentum> sgdMomentumAlgorithm(
        mseObjectiveFunction);

    /* Set input objects for the the Stochastic momentum gradient descent algorithm */
    sgdMomentumAlgorithm.input.set(
        optimization_solver::iterative_solver::inputArgument,
        NumericTablePtr(new HomogenNumericTable<>(initialPoint, 1, nFeatures + 1)));
    sgdMomentumAlgorithm.parameter.learningRateSequence =
        NumericTablePtr(new HomogenNumericTable<>(1, 1, NumericTable::doAllocate, learningRate));
    sgdMomentumAlgorithm.parameter.nIterations = nIterations / 2;
    sgdMomentumAlgorithm.parameter.accuracyThreshold = accuracyThreshold;
    sgdMomentumAlgorithm.parameter.batchSize = batchSize;
    sgdMomentumAlgorithm.parameter.optionalResultRequired = true;

    /* Compute the Stochastic momentum gradient descent result */
    sgdMomentumAlgorithm.compute();

    /* Print computed the Stochastic momentum gradient descent result */
    printNumericTable(
        sgdMomentumAlgorithm.getResult()->get(optimization_solver::iterative_solver::minimum),
        "Minimum after first compute():");
    printNumericTable(
        sgdMomentumAlgorithm.getResult()->get(optimization_solver::iterative_solver::nIterations),
        "Number of iterations performed:");

    /* Set optional result as an optional input */
    sgdMomentumAlgorithm.input.set(
        optimization_solver::iterative_solver::inputArgument,
        sgdMomentumAlgorithm.getResult()->get(optimization_solver::iterative_solver::minimum));
    sgdMomentumAlgorithm.input.set(optimization_solver::iterative_solver::optionalArgument,
                                   sgdMomentumAlgorithm.getResult()->get(
                                       optimization_solver::iterative_solver::optionalResult));

    /* Compute the Stochastic momentum gradient descent result */
    sgdMomentumAlgorithm.compute();

    /* Print computed the Stochastic momentum gradient descent result */
    printNumericTable(
        sgdMomentumAlgorithm.getResult()->get(optimization_solver::iterative_solver::minimum),
        "Minimum after second compute():");
    printNumericTable(
        sgdMomentumAlgorithm.getResult()->get(optimization_solver::iterative_solver::nIterations),
        "Number of iterations performed:");
    return 0;
}
